### PR TITLE
Manipulate masks without scrolling for tablet users

### LINKED ISSUE #10228: [FR] Paint brush mask quick menu.
**Is your feature request related to a problem? Please describe.**
Sometimes when I am creating a mask with the brush I would like to have access to a quick menu where I can modify the brush values without having to constantly press Ctrl, Shift and scroll.

In this quick menu I'll like to have several options like sizing the brush, changing hardness and transparency, as well as viewing, inverting the mask I'm creating, keeping the handles visible and saving the brush parameters as default for later use.

**Describe the solution you'd like**
This is a proposal to use a floating menu to modify the values of the paint brush mask, although I know that some developers don't like floating menus very much.

Using the right click with the mouse a menu would appear where users can modify the brush values, activate the mask display, invert the mask, keep the mask handles visible and if desired, save the brush with these values as default for later use. This would be another way besides using Ctrl or Shift and scroll.

**Alternatives**
Another option could be that when activating the brush, both in the modules and in the mask manager, below the mask manager, this menu is created with these options.

I leave this proposal for your consideration.

**Additional context**

Floating menu with right click.
Brush_settings_00

Other option with menu below the mask manager.
Brush_settings_00a

Option with active icons of view handles, invert and display mask.
Brush_settings_00b

Theme darktable-icons-darker.
Brush_settings_01

Theme darktable-elegant-grey.
Brush_settings_02

Floating menu.
Brush_settings_03

### LINKED ISSUE #11964: Selecting items in lists (treeviews) stops shortcuts working
If you select a list item in a module (for example, selecting an item in the collections module, choosing a mask in the mask manager module), darktable is no longer responsive to keyboard shortcuts until you focus another module or click the centre image. If a key combination is not handled by the list it should probably be passed to the shortcut system for processing.

Not sure if this will have adverse consequences though.

darktable 3.9.0+1673~g30ffb3b5d on ArchLinux

### PR DESCRIPTION
**Pull request by dterrahe**
A frequent complaint has been that to manipulate mask shapes, one often has to use the scroll wheel (to change size, feathering or opacity) which can be inconvenient for pen users. This PR tries to address this somewhat. It consists of 2 main changes and a bit of code cleanup (of which _much_ more would be useful).

1. Add more handles/anchors to circles and ellipses that can be picked up to change size or feather (consistent with paths).
image
image

2. Add a collapsible section to the mask managers with sliders to change the proporties of currently selected (or created) shapes.

image

If a group of shapes has been selected, the soft limits of the sliders are attempting to prevent irreversible distortions (where some of the shapes are clamped at their extreme values whereas others are still adjusted, so that reversing the move doesn't lead to the original situation). But like any soft limits these can be forced.

@TurboGit I would say the main challenge in this PR was the incredible amount of code duplication in the masks area, which makes it very hard to see what's going on. I've cleaned up some, but it is tedious, careful, work that easily breaks stuff if copied code was slightly amended in one location but not in another either accidentally or for good reasons. Once duplicated code blocks exist, they tend to multiply since they are a sign of overdue refactoring. This PR definitely adds somewhat to the problem in that there is a large overlap between what the `modify_property` methods do and what happens in `mouse_scrolled`. This should be solved by having the latter call the former, but I did not want to do this in the initial release because it might break existing functionality while the new has not been properly tested yet. So I hope it doesn't sound to disingenuous to argue that we should be much more reticent to allow duplicated code, especially when large chunks are copied for just a tiny extra functionality. The expectation that "someone" will clean it up hardly ever comes true. Rather, once the duplicates have been duplicated a few times themselves, it massively discourages anyone to touch such toxic code again. I would suggest that if you are presented with a PR that is mostly duplicated code and where the submitter doesn't have the skill to properly refactor, you try to enlist the help of someone knowledgeable in the area. I can point at what happened with #11905 as an example. Yes, it will slow down merging, but PRs like that are hardly ever urgent and it really saves a lot of time, pain and discouragement later on.

Back to masks; apart from just more refactoring, some redesign of the code would be useful too. If, for example, a more complete object would be set up when creating a new shape, we would be able to read properties from the objects themselves, the same as when dealing with existing shapes, rather than in one case using the object and in another having to get values using `dt_conf_get_float`. This would simplify the logic and the code. Again, I haven't made such deep changes yet because I don't want to break the existing workflow while expecting people to test the new functionality. But also because I'm not sure about the lifetime of newly created shapes anyway (especially when the creation process is cancelled by right-clicking; it seem to me the skeletal shape does not get deleted, i.e. that there is a small memory leak).

This PR was obviously inspired by #9015. I tried the path of adding sliders to the top toolbar but felt it led to too many layouting issues and the mask manager was a more appropriate and convenient place. Before that decision I did fix a few small issues plaguing sliders in the middle top location.

This (partially) fixes #10228 (which, tbh, I was not aware of/did not remember).

@elstoc https://darktable-org.github.io/dtdocs/en/darkroom/masking-and-blending/masks/drawn/#available-shapes says that scrolling over a circle's border changes the feathering (even when not holding shift) but this hasn't been the case since #1719 (whether that was a good change or not is for everyone to decide for themselves).

Partially fixes #11964; you can now leave a treeview, for example the masks list, by pressing Tab, so that shortcuts work again. What's going on with the Ctrl-P combination specifically in treeviews is still a mystery to me.

Make 360 degree sliders wrap around while scrolling or using shortcuts.

### CODE REVIEW

**Comment by TurboGit** [comment]
@dterrahe : I fully agree with the refactoring part... But to see duplicated code one need to know the code which is duplicated. I know quite some part of dt code but not all. That being said I'll be more careful about this as again I fully agree with you.

The mask code has certainly not been changed since a long time. Maybe the last changes where from me to handle overlapped masks.

I also agree with you that the large clean-up can be done later even after this PR is merged if needed. It is always good to move step by step otherwise a review won't help to detect possible issues and we will have to rely on testing. 

I'll test and review this soon.

**Comment by TurboGit** [comment]
Testing soon... Was now, I was  eager to see all this in action :) The new controls on circle & ellipse are really nice to have.

The collapsible panel is fine and does not take too much space for people not wanting to use this method. So good for this too for me.

A bug if you use the sliders in the collapsible panel you'll have almost every move recorded into the history. After playing with a circle and an ellipse my history had more than 200 entries ! It seems that the culprit is the opacity slide.

I love the way one can change values for a group, nice !

**Comment by TurboGit** [comment]
We need also to add the new nodes for circle and ellipse into account when checking for the closest control over the mouse. 

I cannot select the circle point (the one on the left to change the size of circle):

image

As it is under the ellipse feather:

image

**Comment by dterrahe** [comment]
> Testing soon... Was now

Thanks for the quick review!

Two fixes, if I understood the distance calculation correctly, for the issues you mentioned. Opacity changes created two history items, so they didn't get merged.

**Comment by dterrahe** [comment]
> The collapsible panel is fine and does not take too much space

Goes without saying that they can also be assigned to shortcuts/midi controllers. Shortcutable actions for the "add shape" buttons in the mask manager were added too.

**Comment by TurboGit** [comment]
Both issues are fixed now. Thanks. I'll continue the test/review.

**Comment by dterrahe** [comment]
Changed bauhaus for degree sliders that cover full 360 range to wrap around when reaching the end. This makes behavior of rotation sliders much more convenient and intuitive, but I believe this is also "correct" behavior for "hue" (in colorbalance etc); please correct me if I'm wrong and the behavior should be made more specific. This applies to scrolling and shortcuts, but dragging sliders isn't changed, unless "accelerated" (by holding ctrl or shift to slow down or speed up moves).

I also found (and fixed) a small issue with "forcing" zoomed soft boundaries. They used to be always at least as wide as softmin-softmax, but can now be zoomed even tighter, so "sticky" narrowing of the active bounds (when you're _at_ the current boundary and move, you want it to be tightened again if possible) used to jump to the soft bound.

**Comment by jenshannoschwalm** [comment]
Can't check on a tablet but seems all fine on Loupedeck+ and via standard mouse for me.

**Comment by dterrahe** [comment]
Right click menus in the mask manager were completely broken by this PR; fixed now and also cleaned up some duplication. Plus multi-select works again.

**Comment by kamidon** [comment]
I tried this out and it definitely improves the experience of editing circular and oval masks with a Wacom tablet. I especially like the on-canvas feather adjustment. Adjusting opacity and rotation work fine from the mask manager. 

I would love to have an on-canvas method of adjusting these instead but the mask manager is still a significant improvement. Mapping a keyboard shortcut link (ctrl-alt-shift-o) to one of the mask manager properties like opacity partially makes this available without diving into the mask manager by popping up the right-click value editor on the canvas at the position of the cursor. With the tablet the pen must be within active range when the shortcut is triggered because otherwise when it is brought into range darktable gets an event that dismisses the popup. This may be due to the mouse position warp that occurs in absolute positioning mode, which I use.

I was unsuccessful in mapping the Wacom buttons as shortcuts. I'll post more about this in issue #1966 as I don't think that is specifically related to this PR.

I did experience a few crashes with messages like:
```
(darktable:120761): Gtk-CRITICAL **: 09:21:24.966: gtk_tree_model_get_iter: assertion 'GTK_IS_TREE_MODEL (tree_model)' failed
an error occurred while trying to execute gdb. please check if gdb is installed on your system.
Magick: abort due to signal 11 (SIGSEGV) "Segmentation Fault"...
backtrace written to /run/user/1000/darktable_bt_62VBT1.txt
```

These seemed to happen when I was attempting to edit mask groups and/or removing an mask from a mask group though I haven't been able to create an deterministic reproduction scenario.

**Comment by dterrahe** [comment]
Thanks for your detailed feedback. Much appreciated!

> popping up the right-click value editor on the canvas at the position of the cursor.

Another thing you could try, which might or might not suit you, is to use extended shortcuts with horizontal or vertical moves. The easiest way to try this out, now that you already have set up a shortrcut for opacity, is to switch on fallbacks (tick "enable fallbacks" in the bottom row of the shortcuts tab or dialog). Now, instead of just pressing ctrl+shift+alt+o once, hold it pressed and at the same time move your mouse/pen left/right or up/down, without clicking/touching). This should move the slider, at different speeds.

> I did experience a few crashes

Please verify if you can also trigger these crashes without this PR. That will significantly help in tracking down the cause and fixing it. Again, your help is much appreciated.

**Comment by kamidon** [comment]
> Please verify if you can also trigger these crashes without this PR. That will significantly help in tracking down the cause and fixing it. Again, your help is much appreciated.

I spent some time with both the 4.0.0 version of darktable (which is what I'm regularly using) and with the build from this PR and was not able to reproduce the crashes with either despite repeatedly trying similar sequences of operations to those I was doing when the crashes happened. I spent some time and tracked down how to look at the core files that were generated from the crashes and they all have a backtrace that looks something like this:
```
(gdb)  bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007f887d4e5c33 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at pthread_kill.c:78
#2  0x00007f887d498b46 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007f887d48354b in __GI_abort () at abort.c:100
#4  0x00007f888017bb87 in MagickPanicSignalHandler () from /nix/store/dl7vyrvllncshdv5y12xp185y03wq3p2-graphicsmagick-1.3.38/lib/libGraphicsMagick.so.3
#5  0x00007f888314f533 in _dt_sigsegv_handler () from /nix/store/nqkwd6lryzkq7pqi4w2r3d23v8wg3l61-darktable-4.1.0-pretest/lib/darktable/libdarktable.so
#6  <signal handler called>
#7  0x0000000000000000 in ?? ()
#8  0x00007f87219d9e02 in alsa_poll () from /nix/store/5ls12s67wllvywxn52z9bjs7icwi17n5-portmidi-2.0.3/lib/libportmidi.so.2
#9  0x00007f87219d80a8 in Pm_Read () from /nix/store/5ls12s67wllvywxn52z9bjs7icwi17n5-portmidi-2.0.3/lib/libportmidi.so.2
#10 0x00007f87219e4be1 in poll_midi_devices () from /nix/store/nqkwd6lryzkq7pqi4w2r3d23v8wg3l61-darktable-4.1.0-pretest/lib/darktable/plugins/lighttable/libmidi.so
#11 0x00007f88828b5874 in g_timeout_dispatch (source=0x4769080, callback=<optimized out>, user_data=<optimized out>) at ../glib/gmain.c:4971
#12 0x00007f88828b4d04 in g_main_dispatch (context=0xe9f510) at ../glib/gmain.c:3417
#13 g_main_context_dispatch (context=0xe9f510) at ../glib/gmain.c:4135
#14 0x00007f88828b50a8 in g_main_context_iterate (context=0xe9f510, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/gmain.c:4211
#15 0x00007f88828b5393 in g_main_loop_run (loop=0xc135700) at ../glib/gmain.c:4411
#16 0x00007f88821df633 in gtk_dialog_run (dialog=0x779a880) at ../gtk/gtkdialog.c:1399
#17 0x00007f8721a45b2e in _show_shortcuts_prefs ()
   from /nix/store/nqkwd6lryzkq7pqi4w2r3d23v8wg3l61-darktable-4.1.0-pretest/lib/darktable/plugins/lighttable/libglobal_toolbox.so
#18 0x00007f8881fb4a65 in _gdk_event_emit (event=event@entry=0x509fa90) at ../gdk/gdkevents.c:73
#19 0x00007f888200b9d2 in gdk_event_source_dispatch (source=<optimized out>, callback=<optimized out>, user_data=<optimized out>) at ../gdk/x11/gdkeventsource.c:367
#20 0x00007f88828b4dfb in g_main_dispatch (context=0xe9f510) at ../glib/gmain.c:3417
#21 g_main_context_dispatch (context=0xe9f510) at ../glib/gmain.c:4135
#22 0x00007f88828b50a8 in g_main_context_iterate (context=0xe9f510, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/gmain.c:4211
#23 0x00007f88828b5393 in g_main_loop_run (loop=loop@entry=0x4f09370) at ../glib/gmain.c:4411
#24 0x00007f888227c375 in gtk_main () at ../gtk/gtkmain.c:1329
#25 0x00007f8883292d1d in dt_gui_gtk_run () from /nix/store/nqkwd6lryzkq7pqi4w2r3d23v8wg3l61-darktable-4.1.0-pretest/lib/darktable/libdarktable.so
#26 0x0000000000402090 in main ()
(gdb) :q
```

The fact that midi polling is in this backtrace seems to indicate to me that the problem is unrelated to this PR. I do have some MIDI devices attached to the computer I'm testing on though I have not attempted to use them with darktable.

> Another thing you could try, which might or might not suit you, is to use extended shortcuts with horizontal or vertical moves. 

Thank you for this tip. I tried this out and it does work nicely with the held keyboard shortcut. If I can get my tablet buttons to trigger shortcuts I think I'd have exactly what I was hoping for!

**Comment by dterrahe** [comment]
If you haven't mapped any midi devices to any shortcuts, then the midi idle polling loop should not trigger any actions. It may just be completely coincidental that the imagemagic panic handler kicks in. Or there may be a memory access race between changes made in the shortcuts dialog (which seems to be open) and the polling routine. But that would surprise me since both of these should be running in the gui thread...

If you can reproduce consistently on that machine with the midi devices attached, I'd really want to know. But I haven't seen this yet myself nor heard about it from other midi users.

**Comment by kamidon** [comment]
> If you can reproduce consistently on that machine with the midi devices attached, I'd really want to know. But I haven't seen this yet myself nor heard about it from other midi users.

If I can find a way to make it happen I'll open a specific bug report for it. Thanks for the help!

**Comment by aurelienpierre** [comment]
> Add a collapsible section to the mask managers with sliders to change the proporties of currently selected (or created) shapes.

See Krita for efficient brush controls implementation for tablet users. Controls are in the top bar and can be called contextually from the image surface with a right click (emulated with button click on Wacom). Hiding them into a collapsible section of a collapsible module is still subpar.

**Comment by dterrahe** [comment]
> still subpar.

But better than nothing, right? I'd happily wait another year for the perfect implementation to appear, including fully configurable and movable toolbars like Krita has, but we could also at this point go with something like this and then strip it out again if it happens to be incompatible with a superior PR submitted later.

**Comment by elstoc** [comment]
Moving controls to the top bar was first suggested back in 2019 (https://github.com/darktable-org/darktable/issues/3740) but withered on the vine and got closed. Personally with my monitor being widescreen and most of my images being 3:2 landscape, I'd prefer the top bar not to get any bigger (since it would shrink my visible image quite a bit) and I'm not sure we could reliably fit all those controls on one line. I always have the 3-line hinter bar switched off.

**Comment by difrkaguilar** [comment]
Nice addition to masks shapes. I've a Wacom tablet. I'll like to test this new panel. Thanks a lot to @dterrahe @AlicVB @TurboGit and all others involved in the development of all those great tools.

Hope to some day someone will integrate other handlers to the shapes if this addition don't implies a lot of work or a huge amount of code. 

This is an idea of the handlers.

Peek 2022-10-03 10-58

**Comment by TurboGit** [comment]
I agree, it looks nice but we do have the besier curve for this and I don't feel we need to have a way to change a circle or ellipse to a curve. One just need to click on the curve button and draw it. The conversion won't avoid the final redraw of the curve anyway, so I see nothing to gain here.

And I also agree that the controls are just fine in the collapsible section. We don't have a real top bar at the moment anyway in dt and the `mask manager` seems a fine location with current dt design.

**Comment by difrkaguilar** [comment]
> I don't feel we need to have a way to change a circle or ellipse to a curve.

I'm sorry I didn't know how to explain the handlers. I made the example with a circle, but perhaps it was not the most appropriate. At no time I was interested in making a modification of this type to a circle or ellipse. It was only a demonstration of the handles used to modify a shape, in that case it would have been better to use a besier. The idea is based on the fact that by selecting several handles or the whole shape, you can easily rotate, scale, compress and stretch it. But it is true that this is more complicated to achieve and may not be the most suitable for adjusting masks.

**Comment by TurboGit** [comment]
@difrkaguilar : Thanks for the clarifications!

**Comment by TurboGit** [comment]
Some issues that I did not see last time, strange...
- The masks are always created with 31% opacity on my side, after adjusting the opacity the next shape is still 31% (eventhough the opacity slider on the blend panel says 100%)
- When moving the cursor for size the feather one is also moving
- When moving the cursor for size or feather the shape (and slider) has some jumps and go back and forth sometime (look at the canvas when moving size from say 20% to 100%, you'll see that the shape sometimes goes smaller before going back to the proper size). This is kind of jerky and not linear as expected.
- Last I had a case where the ellipse was duplicated on canvas, somehow with line across it... really weird but I cannot reproduce, will look if I can find a reproducer.

**Comment by dterrahe** [comment]
I think the opacity slider on the blend panel doesn't relate to the individual shapes but to the overall drawn mask. Is what you are seeing a regression or consistent with previous behavior?

The displayed feather is defined in a relative way but is stored in absolute terms. I've tried not to change anything in the "old" way of doing things, but the way parameter ranges are sort-of restricted and how they are presented could use improvement. And then using the same representation on the sliders might behave better. If I start tinkering with that at this point, however, the risk of introducing regressions becomes much bigger. I'd like the "old" way to continue functioning as before while settling down the new option, to reduce the potential uproar.

I thought I'd fixed most jerkiness issues but will try to reproduce again. Changing properties via sliders causes jumpiness because the first small change immediately triggers reprocessing. Some buffering would be nice, but needs a list of modules that have been changed and that need updating.

When a shape is selected in a group as well as outside it, it gets processed multiple times. Somehow seems to cause problems.

**Comment by TurboGit** [comment]
> I thought I'd fixed most jerkiness issues but will try to reproduce again. Changing properties via sliders causes jumpiness because the first small change immediately triggers reprocessing. Some buffering would be nice, but needs a list of modules that have been changed and that need updating.

Yes, for such thing we usually resort on using : 

`++darktable.gui->reset;`
&
`--darktable.gui->reset;`

and then:

`if(darktable.gui->reset) return;`

See exposure for example, but you probably know this.

**Comment by dterrahe** [comment]
> you probably know this.

Yep! This was one of my first clean-up PRs ;-)

But that is really only meant to prevent infinite recursion when updating sliders from a "changed" signal.

I mean something more like this
https://github.com/darktable-org/darktable/blob/3b18fa928761964a374c954c0f8146cde6365444/src/bauhaus/bauhaus.c#L2894

I've already added (in libs/masks.c) `_timeout_show_all_feathers` so could extend that to only call `dt_masks_iop_update` when the user hasn't changed anything for 2 secs (which would then update the image). You already now see this when dragging an anchor point; the changes are only taken into account _after_ you release the mouse button. But of course when changing properties via scrolling (either the slider or hovering over the shape) there is no "releasing the mouse button" so we'd have to use a timeout. `dt_masks_iop_update` needs to be called for each module that was changed, hence why a list of those would need to be maintained. If we only start regenerating the image after we are done changing the shape, the CPUs will not be busy running openmp routines yet and the gui updates will be smoother.

BTW, I've noticed that _dt_lib_masks_t_ has its own version of _gui_reset_. I suspect that this could (and if so, should) be replaced by the global _darktable.gui->reset_ but again, I don't want to pollute this PR with too many risks of regression.

**Comment by dterrahe** [comment]
A bunch of updates/fixes, mostly to keep sliders and shape properties in sync and avoid jumping due to double updates. The local gui_reset variable has been replaced by the global ++/--darktable.gui->reset approach used everywhere else.

The one change with potentially (much) wider impact than just masks, is the change in bauhaus (cca330a2a3d21178938896b9032525d5629cd2a2) where updates while dragging are now processed as soon as the gui is idle. Previously, an interval dependent on the time it takes to recalc the pipe was used. I believe it is not necessary to wait for the pipe to be empty before sending the next update, because as many updates as have been received while the pipe was busy will be processed as soon as the pipe is empty anyway. I had looked at this change before and didn't notice any adverse effects, but please do keep in mind while testing in case you do notice any different/worse behavior while dragging any other (iop) sliders. For this masks work, it is important that slider updates get processed asap, especially during the creation of a new shape, because then no pipe processing will be triggered anyway and unnecessary waits are useless and annoying.

**Comment by TurboGit** [comment]
@dterrahe : Lot better now, very usable and no more jerky. I just found a little regression, when creating a brush the shift+scroll to control the hardness is lock at about 51,50%, I cannot set it to more while drawing. As soon as the brush mask is applied one can change the hardness all the way from 0% to 100%.

File: src/develop/masks/brush.c [hunk 1]
```diff
@@ -2981,13 +2883,83 @@ static void _brush_initial_source_pos(const float iwd, const float iht, float *x
...
+          ++*count;
+        }
+        pts_number++;
+      }
+      break;
+    case DT_MASKS_PROPERTY_HARDNESS:
+      if(gui->creation)
+      {
+        float masks_hardness = dt_conf_get_float(DT_MASKS_CONF(form->type, brush, hardness));
+        masks_hardness = MAX(BORDER_MIN, MIN(masks_hardness * ratio, BORDER_MAX));
```
Comment (TurboGit) [inline src/develop/masks/brush.c:2928]: BORDER_MIN -> HARDNESS_MIN
BORDER_MAX -> HARDNESS_MAX

**Review by TurboGit — CHANGES_REQUESTED** [review]
Just spotted that.

File: src/develop/masks/brush.c [hunk 2]
```diff
@@ -2981,13 +2883,83 @@ static void _brush_initial_source_pos(const float iwd, const float iht, float *x
...
+        float masks_border = dt_conf_get_float(DT_MASKS_CONF(form->type, brush, border));
+        masks_border = MAX(BORDER_MIN, MIN(masks_border * ratio, BORDER_MAX));
+        dt_conf_set_float(DT_MASKS_CONF(form->type, brush, border), masks_border);
+
+        *sum += 2.0f * masks_border;
+        *max = fminf(*max, BORDER_MAX / masks_border);
+        *min = fmaxf(*min, BORDER_MIN / masks_border);
+        ++*count;
+      }
+      else
```
Comment (TurboGit) [inline src/develop/masks/brush.c:2908]: I'll put the whole else part around `{ }` for readability. And avoids bugs later.

**Review by TurboGit — COMMENTED** [review]
Some minor improvements proposed.

**Comment by TurboGit** [comment]
I have finally been able to reproduce the strange issue I was talking about in my previous comment. To reproduce:
- On exposure (or any module supporting drawn masks) module
- Select drawn mask blend mode
- Select a circle, place it on canvas
- Select an ellipse, place it on canvas
- Fly over the circle on canvas, click to select it
- Fly over the ellipse on canvas, click to select it
- Go to mask manager, change the size slider by clicking on it - see the strange shape :

Capture d'écran de 2022-10-11 13-03-35

**Comment by TurboGit** [comment]
I have not found any other issue than the one reported above. So when fixed it is ok to be merged. A really nice improvement.

**Comment by dterrahe** [comment]
My first attempt to follow your replication instructions got me in an infinite loop inside this code
https://github.com/darktable-org/darktable/pull/2068/files
Seems that when the first 5 "points" are NaN, _next_ gets set to 1 each time and the loop _continue_ s indefinitely. I'm not sure if I understand this piece of code, so I don't want to "fudge" it to just not get stuck but maybe not do the right thing.

Also wondering if this code is responsible for the issue I sometimes have in an ellipse where in one quadrant (usually/always the lower right) when the mouse is inside the ellipse and very close to a radius to one of the anchor points, it does not get detected as being inside.

**Comment by TurboGit** [comment]
I don't get the infinite loop, so maybe a separate issue. On my side I always get a "strange" ellipse on canvas !

**Comment by dterrahe** [comment]
> I always get a "strange" ellipse on canvas !

Basically selecting any shape that wasn't the first deleted the first but added duplicate/modified/mangled version of the selected one. Fixed second counter in for loop; now working for me.

**Comment by TurboGit** [comment]
@dterrahe : Thanks, I'll test that tomorrow now.

**Comment by dterrahe** [comment]
> I don't get the infinite loop, so maybe a separate issue.

Definitely a separate issue (not a regression here). I linked to the code that hangs (or rather the commit that introduced it). 

When the first 5 points are nan, this condition
https://github.com/darktable-org/darktable/blob/eab9af732d8763dad4bd8512d8424fcdd4e58128/src/develop/masks/masks.c#L2112-L2116
causes a loop without changing _i_ or _next_. Not sure yet how to reliably reproduce that situation (I did it by accident when I wasn't paying much attention), but you can easily verify by reading the code that it won't robustly handle it. Not sure what it is _supposed_ to do in that case though. Someone more knowledgeable in that area may also be able to reduce duplication between `dt_masks_point_in_form_near` and `dt_masks_point_in_form_exact` without breaking it.

**Comment by dterrahe** [comment]
I think it would be convenient to apply a log curve to the sliders for size and feathering so that it is easier to manipulate smaller values. Something like
```
static float _log10_curve(float inval, dt_bauhaus_curve_t dir)
{
  return dir == DT_BAUHAUS_SET ? log10f(inval * 99.99f + 0.01f) / 2.0f
                               : (expf(M_LN10 * inval * 2.0f) - 0.01f) / 99.99f;
}
```
and
```
    if(_masks_properties[i].relative)
      dt_bauhaus_slider_set_curve(d->property[i], _log10_curve);
```
inserted after here
https://github.com/darktable-org/darktable/blob/a9b9868a9ec71ed220c60b675c7066c78a509f3e/src/libs/masks.c#L1742

But the bauhaus code does not yet properly support "curves" _and_ soft boundaries at the same time and maybe the stepsize should be impacted as well. So I'll leave this out for now and see if I decide to add it later.

**Comment by dterrahe** [comment]
Fixed a small issue where the mask property sliders would not retain focus. Because if this moving the slider with the arrow keys would not work and when dragging, if you strayed into the slider above or below, that one sometimes would steal the drag. Dragging into the central area would start drawing actions there.

**Review by TurboGit — APPROVED** [review]
Looks good and works fine on my side now. Again thanks a lot for this fine enhancement.

**Comment by difrkaguilar** [comment]
Hello, sorry if this is not the right section to post this information, I don't know if it would be better to report it here or if a new post should be opened for that.

In the newly created panel to manipulate the masks without using the mouse scroll, when you create a mask with gradient, the curvature slider only scrolls to the positive values, and does not show or have the possibility to scroll to the negative values.

It would be nice to have this function as well.

**screenshot**.

Screenshot from 2022-10-12 18-36-22

**Comment by dterrahe** [comment]
> I don't know if it would be better to report it here

Here everybody concerned is already subscribed so you get a quick fix!

Thanks for testing and providing feedback; much appreciated!

**Comment by dterrahe** [comment]
@axelg-de As my canary in the coalmine, do you feel that slider responsiveness has increased since this PR was merged, especially commit cca330a2a3d21178938896b9032525d5629cd2a2 ? On my machine (with slow GPU) I feel that the improvement is more than I expected from having tried it out earlier. Maybe even worth a mention in the release notes. And I haven't noticed any drawbacks that would require finetuning, but maybe you have?

**Comment by AxelG-DE** [comment]
canary tests are done!
- I did checkout at a3dc7d859, which is one commit earlier  than cca330a
- I moved library.db away due to recent updates
- tested with exposure slider and others
- I checked out at dbc3f4f15 (and built, indeed)
- tested again with exposure
- for both tests I also added heavy modules like difuse&sharpen

My findings are clear I think: **no differences before and after cca330a**

(I spent some efforts to make a screencast, only to learn I messed the settings and merely recoded a fraction of my screen, darned, --> deleted)

**Comment by dterrahe** [comment]
OK, many thanks for the work! Reassuring that at least it didn't cause a regression.

I had hoped that it might have improved the update smoothness of the modules themselves a little bit. With that I mean graphs in the modules, for example in filmic, or any of the sliders that move or have gradients that adjust when other sliders move. I think on my slow machine that is now somewhat better, but that may be complete placebo effect. On a fast machine like yours it may not make a difference at all...

**Comment by AxelG-DE** [comment]
you are very welcome

actually there was for a short moment the scenario, that the slider moved nicely and the exposure reacted instantly but the re-draw of the Histogram had a huge lag. Huuuuuge....  but I couldn't reproduce it anymore....

**Pull request merged**